python/src/aac/lang/references.py

```diff
@@ -21,3 +22,198 @@ def filter_definitions_by_reference(definition_to_filter: Definition) -> bool:
+
+
+def get_reference_target_definitions(reference_field_value: str, language_context: LanguageContext) -> list[Definition]:
+    """
+    Return a boolean indicating if the reference target exists in the defined model.
+
+    Args:
+        reference_field_value (str): The definition that is being referenced
+        language_context (LanguageContext): The language context to search.
+
+    Search the language context for target definitions which satisfy the reference.  A matching definition
+    must have all fields referenced in the selector and must contain fields which satisfy any selectors defined in
+    the reference.
+
+    It is assumed that the reference_field_value is a properly formatted reference primitive field.  If it is not,
+    an empty list will be returned.  You can test this using the is_reference_format_valid() method prior to getting
+    the reference target definitions if you have any uncertainty.
+    """
+
+    # Check that input is not None and is valid
+    if not is_reference_format_valid(reference_field_value)[0]:
+        return []
+
+    # First get the root element and optional selector
+    segments = reference_field_value.split('.')
+
+    # get the root so we can pull the right definitions from the language context
+    root, _, _, _ = _get_reference_segment_content(segments[0])
+
+    # recursively filter through the definitions by filtering out definitions using selectors
+    keepers = _process_segment([], segments, language_context.get_definitions_by_root_key(root), language_context)
+
+    return keepers
+
+
+def _process_segment(prefix: list[str], segments: list[str], definitions: list[Definition], language_context: LanguageContext) -> list[Definition]:
+    """This should process segments recursively."""
+    if len(segments) == 0:
+        # everything has been processed, so return what was found in the previous recursion
+        return definitions
+
+    keepers = []
+    key, selector, selector_field, selector_value = _get_reference_segment_content(segments[0])
+
+    if key in language_context.get_root_keys():
+        # handle root key
+        keepers.extend(_process_root(selector, selector_field, selector_value, definitions))
+
+    else:
+        # handle non-root keys
+        keepers.extend(_process_non_root(key, prefix, selector, selector_field, selector_value, definitions))
+
+    if len(keepers) == 0:
+        return []
+
+    nxt_prefix = prefix
+    nxt_prefix.append(key)
+    return _process_segment(nxt_prefix, segments[1:], keepers, language_context)
+
+
+def _process_root(selector: str, selector_field: str, selector_value: str, definitions: list[Definition]) -> list[Definition]:
+    """Return definitions that match the root selector.  If no selector, return all definitions."""
+    keepers = []
+    if selector:
+        # handle root level selector
+        for definition in definitions:
+            if str(definition.get_top_level_fields()[selector_field]) == str(selector_value):
```

Comment:
The tests indicate that you can get away without casting here. I'm happy to help identify/test some additional scenarios which would demonstrate the need to cast these values.
```suggestion
            if definition.get_top_level_fields().get(selector_field) == selector_value:
```

Reply:
I put that cast in because I ran into some hard to diagnose python errors while I was doing development.  It may have become a non-issue as I worked out some bugs, but I'm still hesitant to make to make the change.  As a compromise, I'll add a new issue to expand the test cases and remove casts if they all pass.  I think the issue may have occurred when the field was an integer.  Now I realize I didn't explicitly create a unit test for that condition.